File: packages/shared/src/schemas/chat.schemas.ts
import { z } from '@hono/zod-openapi';
import { AI_API_PROVIDERS } from './provider-key.schemas';
import { AiSdkOptionsSchema, UnifiedModelSchema } from './gen-ai.schemas';
import  { MessageRoleEnum } from './common.schemas';


export type MessageRole = z.infer<typeof MessageRoleEnum>; // Export the type if needed elsewhere

const baseModelOptionsSchema = z.object({
    model: z.string().optional(),
    max_tokens: z.number().optional(),
    temperature: z.number().optional(),
    top_p: z.number().optional(),
    frequency_penalty: z.number().optional(),
    presence_penalty: z.number().optional(),
    stop: z.union([z.string(), z.array(z.string())]).optional(),
});

export type ModelOptions = z.infer<typeof baseModelOptionsSchema>;

// Base schemas for chat entities
export const ChatSchema = z.object({
    id: z.string().openapi({ example: 'chat_1a2b3c4d' }),
    title: z.string(),
    createdAt: z.string().datetime().openapi({ example: '2024-03-10T10:00:00.000Z' }),
    updatedAt: z.string().datetime().openapi({ example: '2024-03-10T10:05:00.000Z' }),
}).openapi('Chat');

export const ChatMessageSchema = z.object({
    id: z.string().min(1).openapi({ example: 'msg-m1a2b3c4', description: 'Message ID' }),
    chatId: z.string().min(1).openapi({ example: 'chat-a1b2c3d4', description: 'Parent Chat ID' }),
    role: MessageRoleEnum.openapi({ example: 'user', description: 'Role of the message sender' }),
    content: z.string().openapi({ example: 'Hello, world!', description: 'Message content' }),
    createdAt: z.string().datetime().openapi({ example: '2024-01-01T12:00:05.000Z', description: 'Creation timestamp (ISO 8601)' })
}).openapi('ChatMessage');

// Request Parameter Schemas
export const ChatIdParamsSchema = z.object({
    chatId: z.string().min(1).openapi({
        param: { name: 'chatId', in: 'path' },
        example: 'chat_1a2b3c4d',
        description: 'The ID of the chat'
    })
}).openapi('ChatIdParams');

// Request Body Schemas
export const CreateChatBodySchema = z.object({
    title: z.string().min(1).openapi({ example: 'New Chat Session' }),
    copyExisting: z.boolean().optional().openapi({ description: 'Copy messages from currentChatId if true' }),
    currentChatId: z.string().min(1).optional().openapi({ example: 'chat-a1b2c3d4' })
}).openapi('CreateChatRequestBody');

export const UpdateChatBodySchema = z.object({
    title: z.string().min(1).openapi({ example: 'Updated Chat Title' }),
}).openapi('UpdateChatRequestBody');

export const CreateChatMessageBodySchema = z.object({
    role: z.string().openapi({ example: 'user', description: 'Message role (user, assistant, system)' }),
    content: z.string().min(1).openapi({ example: 'How can I implement authentication?' }),
}).openapi('CreateChatMessageRequestBody');

// Response Schemas
export const ChatResponseSchema = z.object({
    success: z.literal(true),
    data: ChatSchema
}).openapi('ChatResponse');

export const ChatListResponseSchema = z.object({
    success: z.literal(true),
    data: z.array(ChatSchema)
}).openapi('ChatListResponse');

export const ChatMessageResponseSchema = z.object({
    success: z.literal(true),
    data: ChatMessageSchema
}).openapi('ChatMessageResponse');

export const ChatMessagesListResponseSchema = z.object({
    success: z.literal(true),
    data: z.array(ChatMessageSchema)
}).openapi('ChatMessagesListResponse');


export const MessageListResponseSchema = z.object({
    success: z.literal(true),
    data: z.array(ChatMessageSchema)
}).openapi('MessageListResponse');


export const ModelListResponseSchema = z.object({
    success: z.literal(true),
    data: z.array(UnifiedModelSchema)
}).openapi('ModelListResponse');


export const GetMessagesParamsSchema = z.object({
    chatId: z.string().min(1).openapi({
        param: { name: 'chatId', in: 'path' },
        example: 'chat-a1b2c3d4',
        description: 'The ID of the chat to retrieve messages for'
    })
}).openapi('GetMessagesParams');


export const ForkChatParamsSchema = z.object({
    chatId: z.string().min(1).openapi({
        param: { name: 'chatId', in: 'path' },
        example: 'chat-a1b2c3d4',
        description: 'The ID of the chat to fork'
    })
}).openapi('ForkChatParams');


export const ForkChatBodySchema = z.object({
    excludedMessageIds: z.array(z.string().min(1)).default([]).openapi({
        description: 'Optional list of message IDs to exclude from the fork',
        example: ['msg-m1a2b3c4']
    })
}).openapi('ForkChatRequestBody');

// --- UPDATED: ForkChatFromMessageParamsSchema ---
export const ForkChatFromMessageParamsSchema = z.object({
    chatId: z.string().min(1).openapi({
        param: { name: 'chatId', in: 'path' },
        example: 'chat-a1b2c3d4',
        description: 'The ID of the chat to fork'
    }),
    messageId: z.string().min(1).openapi({
        param: { name: 'messageId', in: 'path' },
        example: 'msg-m1a2b3c4',
        description: 'The ID of the message to fork from'
    })
}).openapi('ForkChatFromMessageParams');


export const ForkChatFromMessageBodySchema = z.object({
    excludedMessageIds: z.array(z.string().min(1)).default([]).openapi({
        description: 'Optional list of message IDs to exclude from the fork',
        example: ['msg-m1a2b3c4']
    })
}).openapi('ForkChatFromMessageRequestBody');


export const UpdateChatParamsSchema = z.object({
    chatId: z.string().min(1).openapi({
        param: { name: 'chatId', in: 'path' },
        example: 'chat-a1b2c3d4',
        description: 'The ID of the chat to update'
    })
}).openapi('UpdateChatParams');

export const DeleteChatParamsSchema = z.object({
    chatId: z.string().min(1).openapi({
        param: { name: 'chatId', in: 'path' },
        example: 'chat-a1b2c3d4',
        description: 'The ID of the chat to delete'
    })
}).openapi('DeleteChatParams');

export const DeleteMessageParamsSchema = z.object({
    messageId: z.string().min(1).openapi({
        param: { name: 'messageId', in: 'path' },
        example: 'msg-m1a2b3c4',
        description: 'The ID of the message to delete'
    })
}).openapi('DeleteMessageParams');


export const ModelsQuerySchema = z.object({
    provider: z.string().openapi({
        description: 'The provider to filter models by',
        example: 'openai'
    })
}).openapi('ModelsQuery');

// --- Schema for individual messages (aligns with Vercel AI SDK CoreMessage) ---
// --- Schema for individual messages (aligns with Vercel AI SDK CoreMessage) ---
export const messageSchema = z.object({
    role: MessageRoleEnum,
    content: z.string(),
    // Keep optional fields if needed internally, but often not needed for basic requests
    // id: z.string().optional(),
    // name: z.string().optional(),
    // tool_call_id: z.string().optional(),
}).openapi('AiMessage');


// --- REVISED: Schema for Streaming Chat Request ---
// Renamed for clarity and modified fields
export const AiChatStreamRequestSchema = z.object({
    chatId: z.string().min(1).openapi({
        example: 'chat-a1b2c3d4',
        description: 'Required ID of the chat session to continue.'
    }),
    userMessage: z.string().min(1, { message: "User message cannot be empty." }).openapi({
        description: 'The latest message content from the user.',
        example: 'Thanks! Can you elaborate on the E=mc^2 part?'
    }),
    provider: z.enum(AI_API_PROVIDERS).or(z.string()).openapi({
        example: 'openrouter',
        description: 'The AI provider to use (e.g., openai, openrouter).'
    }),
    model: z.string().min(1).openapi({
        example: 'deepseek/deepseek-chat-v3-0324:free',
        description: 'The model identifier to use.'
    }),
    options: AiSdkOptionsSchema.openapi({
        description: 'Optional parameters for the AI model.'
    }),
    systemMessage: z.string().optional().openapi({ // Allows overriding system message for this turn
        example: 'Respond concisely.',
        description: 'Optional system message override for this specific request.'
    }),
    tempId: z.string().optional().openapi({
        example: 'temp_msg_456',
        description: 'Temporary client-side ID for optimistic UI updates.'
    }),
    debug: z.boolean().optional().openapi({
        example: true,
        description: 'Enable debug mode for detailed logging.'
    }),
    // 'messages' array is removed - history will be fetched using chatId
    // Fields related to SDK native structured output (schema, enumValues, etc.) are removed
    // but could be added back if needed for the streaming endpoint via `options`.
}).openapi('AiChatStreamRequest');

export type CreateMessageBodyGeneric = {
    message: string;
    chatId: string;
    excludedMessageIds?: string[];
    tempId?: string;
} & ModelOptions

// --- Validation Schemas with OpenAPI Enhancements (Keep as is, looks good) ---
export const chatApiValidation = {
    createChat: {
        body: CreateChatBodySchema
    },
    getMessages: {
        params: GetMessagesParamsSchema
    },
    forkChat: {
        params: ForkChatParamsSchema,
        body: ForkChatBodySchema
    },
    forkChatFromMessage: {
        params: ForkChatFromMessageParamsSchema,
        body: ForkChatFromMessageBodySchema
    },
    updateChat: {
        params: UpdateChatParamsSchema,
        body: UpdateChatBodySchema
    },
    deleteChat: {
        params: DeleteChatParamsSchema
    },
    deleteMessage: {
        params: DeleteMessageParamsSchema
    }
} as const;

// Type exports
export type Chat = z.infer<typeof ChatSchema>;
export type ChatMessage = z.infer<typeof ChatMessageSchema>;
export type CreateChatBody = z.infer<typeof CreateChatBodySchema>;
export type UpdateChatBody = z.infer<typeof UpdateChatBodySchema>;
export type CreateChatMessageBody = z.infer<typeof CreateChatMessageBodySchema>;
export type ExtendedChatMessage = ChatMessage & {
    tempId?: string;
}
export type AiChatStreamRequest = z.infer<typeof AiChatStreamRequestSchema>;

File: packages/python_backend/app/schemas/chat_schemas.py
from typing import Optional, List, Dict, Any, Literal, Union
from pydantic import BaseModel, Field, ConfigDict, field_validator
from datetime import datetime
from enum import Enum
from app.core.config import LOW_MODEL_CONFIG as CORE_LOW_MODEL_CONFIG
from app.schemas.common_schemas import MessageRoleEnum

def convert_timestamp_to_ms_int(value: Any) -> Optional[int]:
    if value is None:
        return None
    if isinstance(value, str):
        try:
            dt_obj = datetime.fromisoformat(value.replace('Z', '+00:00'))
            return int(dt_obj.timestamp() * 1000)
        except ValueError:
            raise ValueError(f"Invalid timestamp string format: {value}")
    elif isinstance(value, (int, float)):
        return int(value)
    elif isinstance(value, datetime):
        return int(value.timestamp() * 1000)
    raise TypeError(f"Timestamp must be an ISO string, int, float, datetime, or None, got {type(value)}")

try:
    from .gen_ai_schemas import AiSdkOptions, UnifiedModel, AiMessage
except ImportError:
    class AiSdkOptions(BaseModel):
        model_config = ConfigDict(title="AiSdkOptions", extra="allow")
        pass
    class UnifiedModel(BaseModel):
        id: int
        name: str
        provider: str
        context_length: Optional[int] = None
        model_config = ConfigDict(title="UnifiedModel", extra="allow")
        pass
    class AiMessage(BaseModel):
        role: MessageRoleEnum
        content: str
        model_config = ConfigDict(title="AiMessage")

class ModelOptions(BaseModel):
    model: Optional[str] = None
    max_tokens: Optional[int] = Field(None, alias="maxTokens")
    temperature: Optional[float] = None
    top_p: Optional[float] = Field(None, alias="topP")
    top_k: Optional[int] = Field(None, alias="topK")
    frequency_penalty: Optional[float] = Field(None, alias="frequencyPenalty")
    presence_penalty: Optional[float] = Field(None, alias="presencePenalty")
    stop: Optional[Union[str, List[str]]] = None
    provider: Optional[str] = None
    model_config = ConfigDict(populate_by_name=True)

def _get_default_chat_model_config():
    config_data = CORE_LOW_MODEL_CONFIG.copy()
    return ModelOptions(**config_data)

class Chat(BaseModel):
    id: int = Field(..., example=20384029823848)
    title: str
    created: int = Field(..., example=1678442400000, description="Creation timestamp (Unix ms)")
    updated: int = Field(..., example=1678442700000, description="Last update timestamp (Unix ms)")
    model_config = ConfigDict(title="Chat")

class ChatMessage(BaseModel):
    id: int = Field(..., example=20384029823848, description="Message ID")
    chat_id: int = Field(..., alias="chatId", example=20384029823848, description="Parent Chat ID")
    role: MessageRoleEnum = Field(..., example="user", description="Role of the message sender")
    content: str = Field(..., example="Hello, world!", description="Message content")
    created: int = Field(..., example=1640995205000, description="Creation timestamp (Unix ms)")
    model_config = ConfigDict(populate_by_name=True)

class ChatIdParams(BaseModel):
    chat_id: int = Field(..., alias="chatId", description="The ID of the chat", json_schema_extra={'param': {'name': 'chatId', 'in': 'path'}, 'example': 20384029823848})
    model_config = ConfigDict(title="ChatIdParams", populate_by_name=True)

class CreateChatBody(BaseModel):
    title: str = Field(..., min_length=1, example="New Chat Session")
    copy_existing: Optional[bool] = Field(None, alias="copyExisting", description="Copy messages from currentChatId if true")
    current_chat_id: Optional[int] = Field(None, alias="currentChatId", example=20384029823848)
    model_config = ConfigDict(populate_by_name=True)

class UpdateChatBody(BaseModel):
    title: str = Field(..., min_length=1, example="Updated Chat Title")

class CreateChatMessageBody(BaseModel):
    role: str = Field(..., example="user", description="Message role (user, assistant, system)")
    content: str = Field(..., min_length=1, example="How can I implement authentication?")

class ChatResponse(BaseModel):
    success: Literal[True] = True
    data: Chat

class ChatListResponse(BaseModel):
    success: Literal[True] = True
    data: List[Chat]

class ChatMessageResponse(BaseModel):
    success: Literal[True] = True
    data: ChatMessage

class ChatMessagesListResponse(BaseModel):
    success: Literal[True] = True
    data: List[ChatMessage]

class OperationSuccessResponse(BaseModel):
    success: Literal[True] = True
    detail: Optional[str] = Field(None, example="Operation completed successfully.")

class MessageListResponse(BaseModel):
    success: Literal[True] = True
    data: List[ChatMessage]

class UnifiedModel(BaseModel):
    id: str = Field(..., description="Model identifier")
    name: str = Field(..., description="User-friendly model name")
    provider: str = Field(..., description="Provider ID")
    context_length: Optional[int] = Field(None, alias="contextLength", description="Context window size in tokens")
    model_config = ConfigDict(title="UnifiedModel", populate_by_name=True)

class ModelListResponse(BaseModel):
    success: Literal[True] = True
    data: List[UnifiedModel]

class GetMessagesParams(BaseModel):
    chat_id: int = Field(..., alias="chatId", description="The ID of the chat to retrieve messages for", json_schema_extra={'param': {'name': 'chatId', 'in': 'path'}, 'example': 20384029823848})
    model_config = ConfigDict(populate_by_name=True)

class ForkChatParams(BaseModel):
    chat_id: int = Field(..., alias="chatId", description="The ID of the chat to fork", json_schema_extra={'param': {'name': 'chatId', 'in': 'path'}, 'example': 20384029823848})
    model_config = ConfigDict(populate_by_name=True)

class ForkChatBody(BaseModel):
    excluded_message_ids: List[int] = Field(default=[], alias="excludedMessageIds", description="Optional list of message IDs to exclude from the fork", json_schema_extra={'example': [20384029823848]})
    model_config = ConfigDict(populate_by_name=True)

class ForkChatFromMessageParams(BaseModel):
    chat_id: int = Field(..., alias="chatId", description="The ID of the chat to fork", json_schema_extra={'param': {'name': 'chatId', 'in': 'path'}, 'example': 20384029823848})
    message_id: int = Field(..., alias="messageId", description="The ID of the message to fork from", json_schema_extra={'param': {'name': 'messageId', 'in': 'path'}, 'example': 20384029823849})
    model_config = ConfigDict(populate_by_name=True)

class ForkChatFromMessageBody(BaseModel):
    excluded_message_ids: List[int] = Field(default=[], alias="excludedMessageIds", description="Optional list of message IDs to exclude from the fork", json_schema_extra={'example': [20384029823848]})
    model_config = ConfigDict(populate_by_name=True)

class UpdateChatParams(BaseModel):
    chat_id: int = Field(..., alias="chatId", description="The ID of the chat to update", json_schema_extra={'param': {'name': 'chatId', 'in': 'path'}, 'example': 20384029823848})
    model_config = ConfigDict(title="UpdateChatParams", populate_by_name=True)

class DeleteChatParams(BaseModel):
    chat_id: int = Field(..., alias="chatId", description="The ID of the chat to delete", json_schema_extra={'param': {'name': 'chatId', 'in': 'path'}, 'example': 20384029823848})
    model_config = ConfigDict(title="DeleteChatParams", populate_by_name=True)

class DeleteMessageParams(BaseModel):
    chat_id: int = Field(..., alias="chatId", description="The ID of the chat containing the message", json_schema_extra={'param': {'name': 'chatId', 'in': 'path'}, 'example': 20384029823848})
    message_id: int = Field(..., alias="messageId", description="The ID of the message to delete", json_schema_extra={'param': {'name': 'messageId', 'in': 'path'}, 'example': 20384029823849})
    model_config = ConfigDict(populate_by_name=True)

class ModelsQuery(BaseModel):
    provider: str = Field(..., description="The provider to filter models by", example="openrouter")
    model_config = ConfigDict(title="ModelsQuery")

class AiSdkOptions(BaseModel):
    temperature: Optional[float] = Field(None, ge=0, le=2, description="Controls randomness of output")
    max_tokens: Optional[int] = Field(None, gt=0, alias="maxTokens", description="Maximum tokens to generate")
    top_p: Optional[float] = Field(None, ge=0, le=1, alias="topP", description="Controls diversity via nucleus sampling")
    frequency_penalty: Optional[float] = Field(None, ge=-2, le=2, alias="frequencyPenalty", description="Penalty for token frequency")
    presence_penalty: Optional[float] = Field(None, ge=-2, le=2, alias="presencePenalty", description="Penalty for token presence")
    top_k: Optional[int] = Field(None, gt=0, alias="topK", description="Restricts token choices to top k")
    stop: Optional[Union[str, List[str]]] = Field(None, description="Stop sequences")
    response_format: Optional[Dict[str, Any]] = Field(None, alias="responseFormat", description="Response format specification")
    provider: Optional[str] = Field(None, description="AI provider")
    model: Optional[str] = Field(None, description="Model identifier")
    model_config = ConfigDict(populate_by_name=True)

class AiChatStreamRequest(BaseModel):
    chat_id: int = Field(..., alias="chatId", example=20384029823848, description="Required ID of the chat session to continue.")
    user_message: str = Field(..., min_length=1, alias="userMessage", description="The latest message content from the user.", example="Thanks! Can you elaborate on the E=mc^2 part?")
    options: Optional[AiSdkOptions] = Field(None, description="Optional parameters for the AI model.")
    system_message: Optional[str] = Field(None, alias="systemMessage", example="Respond concisely.", description="Optional system message override for this specific request.")
    temp_id: Optional[int] = Field(None, alias="tempId", example=1677657600000, description="Temporary client-side ID for optimistic UI updates.")
    debug: Optional[bool] = Field(None, example=True, description="Enable debug mode for detailed logging.")
    model_config = ConfigDict(title="AiChatStreamRequest", populate_by_name=True)

class ExtendedChatMessage(BaseModel):
    id: Optional[int] = Field(None, example=20384029823848, description="Message ID")
    chat_id: int = Field(..., alias="chatId", example=20384029823848, description="Parent Chat ID")
    role: MessageRoleEnum = Field(..., example="user", description="Role of the message sender")
    content: str = Field(..., example="Hello, world!", description="Message content")
    created: Optional[int] = Field(None, example=1640995205000, description="Creation timestamp (Unix ms)")
    temp_id: Optional[int] = Field(None, alias="tempId")
    model_config = ConfigDict(title="ExtendedChatMessage", populate_by_name=True)

